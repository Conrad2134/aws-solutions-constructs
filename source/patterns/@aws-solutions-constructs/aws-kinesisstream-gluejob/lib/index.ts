/**
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { CfnJob, CfnJobProps } from '@aws-cdk/aws-glue';
import { Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { Stream, StreamProps } from '@aws-cdk/aws-kinesis';
import { Bucket, BucketProps, CfnBucket } from '@aws-cdk/aws-s3';
import { Construct } from '@aws-cdk/core';
import * as defaults from '@aws-solutions-constructs/core';


export interface KinesisStreamGlueJobProps {
    /**
     * Existing instance of Kineses Data Stream. If not set, it will create an instance
     */
    readonly existingStreamObj?: Stream;
    /**
     * User provided props to override the default props for the Kinesis Stream.
     *
     * @default - Default props are used
     */
    readonly kinesisStreamProps?: StreamProps | any;
    /**
     * User provides props to override the default props for Glue ETL Jobs. This parameter will be ignored if the
     * existingGlueJob parameter is set
     *
     * @default - Default props are used
     */
    readonly glueJobProps?: CfnJobProps;
    /**
     * Existing GlueJob configuration. If not set, it will create the a CfnJob instance using the glueJobProps
     */
    readonly existingGlueJob?: CfnJob;
}

export class KinesisStreamGlueJob extends Construct {
    public readonly kinesisStream: Stream;

    public readonly glueJob: CfnJob;

    constructor(scope: Construct, id: string, props: KinesisStreamGlueJobProps) {
        super(scope, id);

        this.kinesisStream = defaults.buildKinesisStream(this, {
            existingStreamObj: props.existingStreamObj,
            kinesisStreamProps: props.kinesisStreamProps,
        });

        this.glueJob = defaults.buildGlueJob(this, {
            existingCfnJob: props.existingGlueJob,
            glueJobProps: props.glueJobProps
        });

        let _glueJobRoleName: string;
        if (props.existingGlueJob === undefined) {
            _glueJobRoleName = props.glueJobProps?.role!;
        } else {
            _glueJobRoleName = props.existingGlueJob.role;
        }

        this.kinesisStream.grantRead(Role.fromRoleArn(scope, 'GlueJobRole', _glueJobRoleName));
    }

    /**
     * This is a helper method to create the Role required for the Glue Job. If a role is already created then this
     * method is not required to be called.
     *
     * @param scope - The AWS Construct under which the role is to be created
     */
    public static createGlueJobRole(scope: Construct): Role {
        const _jobRole: Role = new Role(scope, 'JobRole', {
            assumedBy: new ServicePrincipal('glue.amazonaws.com'),
            description: 'Service role that Glue custom ETL jobs will assume for exeuction',
        });

        return _jobRole;
    }

    /**
     * This is a helper method to creates @CfnJob.JobCommandProperty for CfnJob. Based on the input parameters provided,
     * it will also create the S3 bucket for the ETL script location and grant 'read' access to 'glue.amazonaws.com'
     * Service Principal so that the script inside the bucket can be accessed as by AWS Glueglobal.fetch = require('node-fetch');
     *
     * Also this method does not set lifecycle policies on S3 buckets created unless they are explicitly set in the bucket
     * props
     *
     * @param scope - The AWS Construct under the underlying construct should be created
     * @param _jobID - The identifier/ name of the ETL Job
     * @param pythonVersion - The values as for Glue Documentation are '2' and '3'. There is no validation in the
     * method to check for these values to be forward compatible with Glue API changes. If valid values are not provided
     * the deployment would fail.
     * @param existingScriptLocation - If an S3 bucket location for the script exists, set this parameter. If the Bucket
     * is to be created, set the value as undefined. Setting this parameter will ignore @scriptLocationBucketProps as the
     * bucket already exists
     * @param scriptLocationBucketProps - Set this parameter only if the bucket is to be created. If the
     * @existingScriptLocation parameter is set, this parameter will be ignored. This parameter allows to set S3 Bucket
     * properts where the ETL script will be located
     */
    public static createGlueJobCommand(scope: Construct,
                                       _jobID: string, pythonVersion: string, existingScriptLocation?: Bucket,
                                       scriptLocationBucketProps?: BucketProps): CfnJob.JobCommandProperty {
        // create s3 bucket where script can be deployed
        let scriptLocation: [ Bucket, (Bucket | undefined)? ];
        if (existingScriptLocation === undefined) {
            scriptLocation = defaults.buildS3Bucket(scope, {
                bucketProps: scriptLocationBucketProps
            });
            // Remove the default LifecycleConfiguration for the Logging Bucket
            if (scriptLocationBucketProps?.lifecycleRules === undefined) {
                (scriptLocation[0]!.node.findChild('Resource') as CfnBucket).addPropertyDeletionOverride(
                                                                                        'LifecycleConfiguration.Rules');
            }
        } else {
            // since bucket location was provided in the props, logger bucket is not created
            scriptLocation = [ existingScriptLocation, undefined ];
        }

        scriptLocation[0].grantRead(new ServicePrincipal('glue.amazonaws.com'));

        const _jobCommand: CfnJob.JobCommandProperty = {
            name: _jobID,
            pythonVersion,
            scriptLocation: scriptLocation[0].s3UrlForObject()
        };

        return _jobCommand;
    }
}