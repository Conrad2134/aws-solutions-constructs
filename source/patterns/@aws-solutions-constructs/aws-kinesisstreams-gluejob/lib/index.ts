/**
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { CfnDatabase, CfnJob, CfnJobProps, CfnTable } from '@aws-cdk/aws-glue';
import { CfnPolicy, Effect, IRole, Policy, PolicyStatement, Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { Stream, StreamProps } from '@aws-cdk/aws-kinesis';
import { Asset } from '@aws-cdk/aws-s3-assets';
import { Aws, Construct } from '@aws-cdk/core';
import * as defaults from '@aws-solutions-constructs/core';

export interface KinesisStreamGlueJobProps {
  /**
   * Existing instance of Kineses Data Stream. If not set, it will create an instance
   */
  readonly existingStreamObj?: Stream;
  /**
   * User provided props to override the default props for the Kinesis Stream.
   *
   * @default - Default props are used
   */
  readonly kinesisStreamProps?: StreamProps | any;
  /**
   * User provides props to override the default props for Glue ETL Jobs. This parameter will be ignored if the
   * existingGlueJob parameter is set
   *
   * @default - Default props are used
   */
  readonly glueJobProps?: CfnJobProps;
  /**
   * Existing GlueJob configuration. If not set, it will create the a CfnJob instance using the glueJobProps
   */
  readonly existingGlueJob?: CfnJob;
  /**
   * A JSON document defining the schema structure of the records in the data stream. An example of such a
   * definition as below. Either @table or @fieldSchema is mandatory. If @table is provided then @fieldSchema is ignored
   * 	"FieldSchema": [{
   *  	"name": "id",
   *  	"type": "int",
   *    "comment": "Identifier for the record"
   *  }, {
   *    "name": "name",
   *    "type": "string",
   *    "comment": "The name of the record"
   *  }, {
   *    "name": "type",
   *    "type": "string",
   *    "comment": "The type of the record"
   *  }, {
   *    "name": "numericvalue",
   *    "type": "int",
   *    "comment": "Some value associated with the record"
   *  },
   */
  readonly fieldSchema?: CfnTable.ColumnProperty [];
  /**
   * Glue Database for this construct. If not provided the construct will create a new Glue Database.
   * The database is where the schema for the data in Kinesis Data Streams
   */
  readonly database?: CfnDatabase;
  /**
   * Glue Table for this construct, If not provided the construct will create a new Table in the
   * database. This table should define the schema for the records in the Kinesis Data Streams.
   * Either @table or @fieldSchema is mandatory. If @table is provided then @fieldSchema is ignored
   */
  readonly table?: CfnTable;
}

export class KinesisStreamGlueJob extends Construct {
  public readonly kinesisStream: Stream;

  public readonly glueJob: CfnJob;

  constructor(scope: Construct, id: string, props: KinesisStreamGlueJobProps) {
    super(scope, id);

    this.kinesisStream = defaults.buildKinesisStream(this, {
      existingStreamObj: props.existingStreamObj,
      kinesisStreamProps: props.kinesisStreamProps,
    });

    this.glueJob = defaults.buildGlueJob(this, {
      existingCfnJob: props.existingGlueJob,
      glueJobProps: props.glueJobProps
    });

    let _glueDatabase: CfnDatabase;
    if (props.database !== undefined) {
      _glueDatabase = props.database!;
    } else {
      _glueDatabase = KinesisStreamGlueJob.createGlueDatabase(scope);
    }

    if (props.fieldSchema === undefined && props.table === undefined) {
      throw Error('Either fieldSchema or table property has to be set, both cannot be optional');
    }

    let _glueTable: CfnTable;
    if (props.table !== undefined) {
      _glueTable = props.table;
    } else {
      _glueTable = KinesisStreamGlueJob.createGlueTable(scope, _glueDatabase, props.fieldSchema!, 'kinesis', {
        STREAM_NAME: this.kinesisStream.streamName
      });
    }

    this.buildRolePolicy(scope, _glueDatabase, _glueTable, this.glueJob.role);
  }

  private buildRolePolicy(scope: Construct, glueDatabase: CfnDatabase, glueTable: CfnTable, role: string): IRole {
    const _glueJobPolicy = new Policy(scope, 'GlueJobPolicy', {
      statements: [ new PolicyStatement({
        effect: Effect.ALLOW,
        actions: [ 'glue:GetJob' ],
        resources: [ `arn:${Aws.PARTITION}:glue:${Aws.REGION}:${Aws.ACCOUNT_ID}:job/${this.glueJob.ref}` ]
      }), new PolicyStatement({
        effect: Effect.ALLOW,
        actions: [ 'glue:GetSecurityConfiguration' ],
        resources: [ '*' ] // Security Configurations have no resource ARNs
      }), new PolicyStatement({
        effect: Effect.ALLOW,
        actions: [ 'glue:GetTable' ],
        resources: [ `arn:${Aws.PARTITION}:glue:${Aws.REGION}:${Aws.ACCOUNT_ID}:table/${glueDatabase.ref}/${glueTable.ref}`,
          `arn:${Aws.PARTITION}:glue:${Aws.REGION}:${Aws.ACCOUNT_ID}:database/${glueDatabase.ref}`,
          `arn:${Aws.PARTITION}:glue:${Aws.REGION}:${Aws.ACCOUNT_ID}:catalog`
        ]
      }), new PolicyStatement({
        effect: Effect.ALLOW,
        actions: [ 'cloudwatch:PutMetricData' ],
        resources: [ '*' ], // Metrics do not have resource ARN and hence added conditions
        conditions: {
          StringEquals: {
            "cloudwatch:namespace": "AWS/Glue"
          }
        }
      }), new PolicyStatement({
        effect: Effect.ALLOW,
        actions: [ 'kinesis:DescribeStream', 'kinesis:DescribeStreamSummary', 'kinesis:GetRecords',
          'kinesis:GetShardIterator', 'kinesis:ListShards', 'kinesis:SubscribeToShard' ],
        resources: [ this.kinesisStream.streamArn ]
      })]
    });

    (_glueJobPolicy.node.defaultChild as CfnPolicy).cfnOptions.metadata = {
      cfn_nag: {
        rules_to_suppress: [{
          id: 'W12',
          reason: 'Glue Security Configuration does not have an ARN, and the policy only allows reading the configuration.\
            CloudWatch metrics also do not have an ARN but adding a namespace condition to the policy to allow it to\
            publish metrics only for AWS Glue'
        }]
      }
    };

    const _glueJobRole = Role.fromRoleArn(scope, 'GlueJobRole', role);
    _glueJobRole.attachInlinePolicy(_glueJobPolicy);
    return _glueJobRole;
  }

  /**
   * This is a helper method to create the Role required for the Glue Job. If a role is already created then this
   * method is not required to be called.
   *
   * @param scope - The AWS Construct under which the role is to be created
   */
  public static createGlueJobRole(scope: Construct): Role {
    return new Role(scope, 'JobRole', {
      assumedBy: new ServicePrincipal('glue.amazonaws.com'),
      description: 'Service role that Glue custom ETL jobs will assume for exeuction',
    });
  }

  /**
   * This is a helper method to creates @CfnJob.JobCommandProperty for CfnJob. Based on the input parameters, If the
   * @S3ObjectUrlForScript is passed, it will create only the JobCommandProperty. Instead if the @scriptLocationPath is
   * passed it will an Asset from the @scriptLocationPath and returns the JobCommandProperty and the Asset. The script
   * location can be retrieved using @Asset.s3ObjectUrl
   *
   * @param scope - The AWS Construct under the underlying construct should be created
   * @param _commandName - The identifier/ name of the ETL Job. The values are glueetl, gluestreaming, pythonshell.
   * THere is no validation, but if valid values are not provided, the deployment may fail
   * @param pythonVersion - The values as for Glue Documentation are '2' and '3'. There is no validation in the
   * method to check for these values to be forward compatible with Glue API changes. If valid values are not provided
   * the deployment may fail.
   * @param s3ObjectUrlForScript - If an S3 bucket location for the script exists, set this parameter. If the Bucket
   * is to be created, set the value as undefined. Setting this parameter will ignore @scriptLocationBucketProps as the
   * bucket already exists
   * @param scriptLocationPath - Set this parameter only if the bucket is to be created. If not then a new
   * Bucket location will be created to upload the ETL script asset
   */
  public static createGlueJobCommand(scope: Construct, _commandName: string, pythonVersion: string, glueJobRole: Role,
    s3ObjectUrlForScript?: string, scriptLocationPath?: string): [ CfnJob.JobCommandProperty, Asset? ] {
    // create s3 bucket where script can be deployed
    let _scriptLocation: string;
    let _assetLocation: Asset;
    if (s3ObjectUrlForScript === undefined) {
      if (scriptLocationPath === undefined) {
        throw Error('Either s3ObjectUrlForScript or scriptLocationPath is required');
      } else {
        _assetLocation = new Asset(scope, 'ETLScriptLocation', {
          path: scriptLocationPath!
        });
        _assetLocation.grantRead(glueJobRole);
        _scriptLocation = _assetLocation.s3ObjectUrl;
      }
    } else {
      // since bucket location was provided in the props, logger bucket is not created
      _scriptLocation = s3ObjectUrlForScript;
    }

    return [{
      name: _commandName,
      pythonVersion,
      scriptLocation: _scriptLocation
    }, _assetLocation! !== undefined ? _assetLocation! : undefined ];
  }

  public static createGlueTable(scope: Construct, database: CfnDatabase, fieldSchema: CfnTable.ColumnProperty [], sourceType: string,
    parameters?: any): CfnTable {
    return defaults.DefaultGlueTable(scope, database, fieldSchema, sourceType, parameters);
  }

  public static createGlueDatabase(scope: Construct): CfnDatabase {
    return defaults.DefaultGlueDatabase(scope);
  }
}