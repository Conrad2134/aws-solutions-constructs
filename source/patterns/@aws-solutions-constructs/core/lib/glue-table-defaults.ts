/**
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { CfnDatabase, CfnTable } from "@aws-cdk/aws-glue";
import { Aws, Construct } from "@aws-cdk/core";

/**
 *
 * @param scope
 * @param database
 * @param fieldSchema
 * @param sourceType
 * @param parameters - Key value pairs of parmaeters. If source type is 'Kinesis', pass Kinesis Data Stream name with key
 *  as 'STREAM_NAME'. Example: {STREAM_NAME: 'KinesisStreamConstrct.streamName'}
 */
export function DefaultGlueTableProps(scope: Construct, database: CfnDatabase, fieldSchema: CfnTable.ColumnProperty [],
                                      sourceType: string, parameters?: any): CfnTable {
  let _glueTable: CfnTable;
  if (sourceType === 'Kinesis') {
    const kinesisStreamName = parameters.STREAM_NAME;
    _glueTable = new CfnTable(scope, 'GlueTable', {
      catalogId: database.catalogId,
      databaseName: database.ref,
      tableInput: {
        storageDescriptor: {
          columns: fieldSchema,
          location: kinesisStreamName,
          inputFormat: "org.apache.hadoop.mapred.TextInputFormat",
          outputFormat: "org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat",
          compressed: false,
          numberOfBuckets: -1,
          serdeInfo: {
            serializationLibrary: "org.openx.data.jsonserde.JsonSerDe",
            parameters: {
              paths: '.'
            }
          },
          parameters: {
            endpointUrl: `https://kinesis.${Aws.REGION}.amazonaws.com`,
            streamName: kinesisStreamName,
            typeOfData: "kinesis"
          }
        },
        tableType: 'EXTERNAL_TABLE',
        parameters: {
          classication: 'json'
        }
      }
    });
  } else {
    throw Error('Source Type not Supported. Valid Source Type not provided');
  }

  return _glueTable!;
}