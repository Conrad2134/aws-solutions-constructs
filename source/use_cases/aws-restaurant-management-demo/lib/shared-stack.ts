/**
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

// Imports
import * as cdk from '@aws-cdk/core';
import * as dynamodb from '@aws-cdk/aws-dynamodb';

// Stack
export class SharedStack extends cdk.Stack {

  // Public variables
  public readonly database: dynamodb.Table;

  // Constructor
  constructor(scope: cdk.Construct, id: string, props: cdk.StackProps) {
    super(scope, id, props);

    // Database for storing orders
    this.database = new dynamodb.Table(this, 'database', {
      partitionKey: {
        name: 'id',
        type: dynamodb.AttributeType.STRING
      }
    });
  }
}