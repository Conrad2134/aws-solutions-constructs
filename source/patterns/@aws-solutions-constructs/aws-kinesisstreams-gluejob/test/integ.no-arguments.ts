/**
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

// Imports
import { CfnJob } from '@aws-cdk/aws-glue';
import { Bucket } from '@aws-cdk/aws-s3';
import { App, Duration, Stack } from '@aws-cdk/core';
import { KinesisStreamGlueJob, KinesisStreamGlueJobProps } from '../lib';

// Setup
const app = new App();
const stack = new Stack(app, 'test-kinesisstream-gluejob');
stack.templateOptions.description = 'Integration Test for aws-kinesisstream-gluejob';

const _jobCommand: CfnJob.JobCommandProperty = {
  name: 'glueetl',
  pythonVersion: '3',
  scriptLocation: new Bucket(stack, 'existingScriptLocation', {
    versioned: false,
    lifecycleRules: [{
      expiration: Duration.days(30)
    }]
  }).s3UrlForObject()
};

// Definitions
const props: KinesisStreamGlueJobProps = {
  glueJobProps: {
    command: _jobCommand,
  },
  fieldSchema: [{
    name: "id",
    type: "int",
    comment: "Identifier for the record"
  }, {
    name: "name",
    type: "string",
    comment: "The name of the record"
  }, {
    name: "type",
    type: "string",
    comment: "The type of the record"
  }, {
    name: "numericvalue",
    type: "int",
    comment: "Some value associated with the record"
  }],
};

new KinesisStreamGlueJob(stack, 'test-kinesisstreams-lambda', props);

// Synth
app.synth();