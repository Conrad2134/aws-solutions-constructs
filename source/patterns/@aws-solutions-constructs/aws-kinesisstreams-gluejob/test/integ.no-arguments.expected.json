{
  "Description": "Integration Test for aws-kinesisstream-gluejob",
  "Resources": {
    "JobRole014917C6": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "glue.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "Description": "Service role that Glue custom ETL jobs will assume for exeuction"
      }
    },
    "JobRoleDefaultPolicy5DE0D8F9": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*"
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":s3:::",
                      {
                        "Ref": "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3BucketCCACFC1B"
                      }
                    ]
                  ]
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":s3:::",
                      {
                        "Ref": "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3BucketCCACFC1B"
                      },
                      "/*"
                    ]
                  ]
                }
              ]
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "JobRoleDefaultPolicy5DE0D8F9",
        "Roles": [
          {
            "Ref": "JobRole014917C6"
          }
        ]
      }
    },
    "testkinesisstreamslambdaKinesisStream374D6D56": {
      "Type": "AWS::Kinesis::Stream",
      "Properties": {
        "ShardCount": 1,
        "RetentionPeriodHours": 24,
        "StreamEncryption": {
          "EncryptionType": "KMS",
          "KeyId": "alias/aws/kinesis"
        }
      }
    },
    "testkinesisstreamslambdaLogPolicy5FB58427": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents"
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition"
                    },
                    ":logs:",
                    {
                      "Ref": "AWS::Region"
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId"
                    },
                    ":log-group:/aws-glue/*"
                  ]
                ]
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "testkinesisstreamslambdaLogPolicy5FB58427",
        "Roles": [
          {
            "Fn::Select": [
              1,
              {
                "Fn::Split": [
                  "/",
                  {
                    "Fn::Select": [
                      5,
                      {
                        "Fn::Split": [
                          ":",
                          {
                            "Fn::GetAtt": [
                              "JobRole014917C6",
                              "Arn"
                            ]
                          }
                        ]
                      }
                    ]
                  }
                ]
              }
            ]
          }
        ]
      }
    },
    "testkinesisstreamslambdaETLJob44B50676": {
      "Type": "AWS::Glue::Job",
      "Properties": {
        "Command": {
          "Name": "pythonshell",
          "PythonVersion": "3",
          "ScriptLocation": {
            "Fn::Join": [
              "",
              [
                "s3://",
                {
                  "Ref": "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3BucketCCACFC1B"
                },
                "/",
                {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3VersionKey753DB1DE"
                        }
                      ]
                    }
                  ]
                },
                {
                  "Fn::Select": [
                    1,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3VersionKey753DB1DE"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        "Role": {
          "Fn::GetAtt": [
            "JobRole014917C6",
            "Arn"
          ]
        },
        "SecurityConfiguration": "testSecConfig"
      }
    },
    "GlueDatabase": {
      "Type": "AWS::Glue::Database",
      "Properties": {
        "CatalogId": {
          "Ref": "AWS::AccountId"
        },
        "DatabaseInput": {
          "Description": "A Glue database table"
        }
      }
    },
    "GlueTable": {
      "Type": "AWS::Glue::Table",
      "Properties": {
        "CatalogId": {
          "Ref": "AWS::AccountId"
        },
        "DatabaseName": {
          "Ref": "GlueDatabase"
        },
        "TableInput": {
          "Parameters": {
            "classication": "json"
          },
          "StorageDescriptor": {
            "Columns": [
              {
                "Comment": "Identifier for the record",
                "Name": "id",
                "Type": "int"
              },
              {
                "Comment": "The name of the record",
                "Name": "name",
                "Type": "string"
              },
              {
                "Comment": "The type of the record",
                "Name": "type",
                "Type": "string"
              },
              {
                "Comment": "Some value associated with the record",
                "Name": "numericvalue",
                "Type": "int"
              }
            ],
            "Compressed": false,
            "InputFormat": "org.apache.hadoop.mapred.TextInputFormat",
            "Location": {
              "Ref": "testkinesisstreamslambdaKinesisStream374D6D56"
            },
            "NumberOfBuckets": -1,
            "OutputFormat": "org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat",
            "Parameters": {
              "endpointUrl": {
                "Fn::Join": [
                  "",
                  [
                    "https://kinesis.",
                    {
                      "Ref": "AWS::Region"
                    },
                    ".amazonaws.com"
                  ]
                ]
              },
              "streamName": {
                "Ref": "testkinesisstreamslambdaKinesisStream374D6D56"
              },
              "typeOfData": "kinesis"
            },
            "SerdeInfo": {
              "Parameters": {
                "paths": "."
              },
              "SerializationLibrary": "org.openx.data.jsonserde.JsonSerDe"
            }
          },
          "TableType": "EXTERNAL_TABLE"
        }
      }
    },
    "GlueJobRolePolicy422D4F2E": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "kinesis:DescribeStreamSummary",
                "kinesis:GetRecords",
                "kinesis:GetShardIterator",
                "kinesis:ListShards",
                "kinesis:SubscribeToShard"
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "testkinesisstreamslambdaKinesisStream374D6D56",
                  "Arn"
                ]
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "GlueJobRolePolicy422D4F2E",
        "Roles": [
          {
            "Fn::Select": [
              1,
              {
                "Fn::Split": [
                  "/",
                  {
                    "Fn::Select": [
                      5,
                      {
                        "Fn::Split": [
                          ":",
                          {
                            "Fn::GetAtt": [
                              "JobRole014917C6",
                              "Arn"
                            ]
                          }
                        ]
                      }
                    ]
                  }
                ]
              }
            ]
          }
        ]
      }
    },
    "GlueJobPolicyAEA4B94E": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "glue:GetJob",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition"
                    },
                    ":glue:",
                    {
                      "Ref": "AWS::Region"
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId"
                    },
                    ":job/",
                    {
                      "Ref": "testkinesisstreamslambdaETLJob44B50676"
                    }
                  ]
                ]
              }
            },
            {
              "Action": "glue:GetSecurityConfiguration",
              "Effect": "Allow",
              "Resource": "*"
            },
            {
              "Action": "glue:GetTable",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":glue:",
                      {
                        "Ref": "AWS::Region"
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId"
                      },
                      ":table/",
                      {
                        "Ref": "GlueDatabase"
                      },
                      "/",
                      {
                        "Ref": "GlueTable"
                      }
                    ]
                  ]
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":glue:",
                      {
                        "Ref": "AWS::Region"
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId"
                      },
                      ":database/",
                      {
                        "Ref": "GlueDatabase"
                      }
                    ]
                  ]
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":glue:",
                      {
                        "Ref": "AWS::Region"
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId"
                      },
                      ":catalog"
                    ]
                  ]
                }
              ]
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "GlueJobPolicyAEA4B94E",
        "Roles": [
          {
            "Fn::Select": [
              1,
              {
                "Fn::Split": [
                  "/",
                  {
                    "Fn::Select": [
                      5,
                      {
                        "Fn::Split": [
                          ":",
                          {
                            "Fn::GetAtt": [
                              "JobRole014917C6",
                              "Arn"
                            ]
                          }
                        ]
                      }
                    ]
                  }
                ]
              }
            ]
          }
        ]
      }
    }
  },
  "Parameters": {
    "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3BucketCCACFC1B": {
      "Type": "String",
      "Description": "S3 bucket for asset \"13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303ac\""
    },
    "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3VersionKey753DB1DE": {
      "Type": "String",
      "Description": "S3 key for asset version \"13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303ac\""
    },
    "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acArtifactHash2A1C2D40": {
      "Type": "String",
      "Description": "Artifact hash for asset \"13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303ac\""
    }
  },
  "Outputs": {
    "ScriptLocation": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "s3://",
            {
              "Ref": "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3BucketCCACFC1B"
            },
            "/",
            {
              "Fn::Select": [
                0,
                {
                  "Fn::Split": [
                    "||",
                    {
                      "Ref": "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3VersionKey753DB1DE"
                    }
                  ]
                }
              ]
            },
            {
              "Fn::Select": [
                1,
                {
                  "Fn::Split": [
                    "||",
                    {
                      "Ref": "AssetParameters13d7b7f7fba24472b473a111e804308f745de51605e9730c69700262c57303acS3VersionKey753DB1DE"
                    }
                  ]
                }
              ]
            }
          ]
        ]
      }
    }
  }
}